consensus/istanbul/backend/engine.go

```diff
@@ -95,15 +96,68 @@ var (
 	nonceAuthVote = hexutil.MustDecode("0xffffffffffffffff") // Magic nonce number to vote on adding a new validator
 	nonceDropVote = hexutil.MustDecode("0x0000000000000000") // Magic nonce number to vote on removing a validator.
 
-	inmemoryAddresses  = 20 // Number of recent addresses from ecrecover
-	recentAddresses, _ = lru.NewARC(inmemoryAddresses)
+	inmemoryBlocks             = 100 // Number of blocks to precompute validators' addresses
+	inmemoryValidatorsPerBlock = 30  // Approximate number of validators' addresses from ecrecover
+	recentAddresses, _         = lru.NewARC(inmemoryBlocks)
+	signatureAddresses, _      = lru.New(inmemoryBlocks * inmemoryValidatorsPerBlock)
 )
 
 // Author retrieves the Klaytn address of the account that minted the given block.
 func (sb *backend) Author(header *types.Header) (common.Address, error) {
 	return ecrecover(header)
 }
 
+// CanVerifyHeadersConcurrently returns true if concurrent header verification possible, otherwise returns false.
+func (sb *backend) CanVerifyHeadersConcurrently() bool {
+	return false
+}
+
+// PreprocessHeaderVerification prepares header verification for heavy computation before synchronous header verification such as ecrecover.
+func (sb *backend) PreprocessHeaderVerification(headers []*types.Header) (chan<- struct{}, <-chan error) {
+	abort := make(chan struct{})
+	results := make(chan error, inmemoryBlocks)
```

Comment:
Why did you use `inmemoryBlocks` instead of `len(headers)`?

Reply:
Is 100 * 30 enough cache size? Did downloader send thousands of blocks at one time?

Reply:
> Why did you use inmemoryBlocks instead of len(headers)?

This is for throttling purpose. If the computing is way faster and it replaces the already computed ones from cache, this precomputing is not useful. It may use more memory, but I think it is not necessary.

> Is 100 * 30 enough cache size? Did downloader send thousands of blocks at one time?

So, the throttling header verification every 100 blocks, and each block may contain 30 signatures at most. So, I put that number.

Reply:
I mean throttling 100 blocks doesn't seem to be enough. Downloader could send 2048 blocks at once, but it only preprocesses 100 blocks.

Reply:
@aidan-kwon I will increase the number to 2048. I thought the cache is not meaningful after syncing, so the space would be wasted (although only about 4MB is used).

Reply:
Thanks. I think `len(headers)` is more suitable for the length of `results` channel, though. `inmemoryBlocks` looks fine also.